ONSdigital/rasrm-acceptance-tests
# Standalone acceptance tests

# Motivation and Context
Some Acceptance tests rely on data state set by previous tests which, not only prevents them from running independently but just changing the feature file name, can cause them to fail.

The goal was to establish a pattern whereby each Scenario runs in isolation.

A "@standlone" tag differentiates between standalone/non-standalone Scenarios. Non-standalone (or sequential) tests should run exactly as before but "@standlone" now indicates a Scenario can run in parallel and optionally, create its own data using a "@fixture.xxxxxxx" tag.

About 18% of Scenarios have been converted so far - 

	add_a_survey.feature (6 scenarios)
	breadcrumb_navigation.feature  (3 scenarios)
	display_unused_enrolment_code.feature (1 scenario)
	generate_new_enrolment_code.feature (1 scenario)
	sign_in_internal.feature (7 scenarios)
	sign_out_internal.feature (1 scenario)
	social_change_case_status.feature (3 scenarios)
	social_disable_uac.feature (1 scenario)
	social_find_case_by_postcode.feature (2 scenarios)
	social_sign_in_internal.feature (7 scenarios)
	social_sign_out_internal.feature (1 scenario)
	social_view_case_details.feature (1 scenario)
	survey_enrolment.feature (2 scenarios)
	view_collection_exercise_created_state.feature (2 scenarios)

Scenarios that use the @fixture tag, use data from their Feature/Scenario to improve traceability - 

	Survey Data
	=========
	Survey name = 'BUSINESS' OR 'SOCIAL' + "-" + feature name + UTC creation timestamp (may be compacted)
	Short name = RU Reference (unique per Scenario)
	Survey Ref  = a randomly generated number above 90000 (keeps standlone tests together on "View list of business surveys" page)

	Collection Exercise Data
	==================
	User Description = 'BUSINESS' OR 'SOCIAL' + "-" + scenario name (may be compacted)
	A Respondent login (if required) =  generated from RU Reference e.g 53621951874@53621951874.com

        A typical @fixture tag test takes around 30 seconds to run, in parallel, this reduces to about 8 seconds 

Negatives
=========
Time to convert the remaining 170+ Scenarios
New acceptance tests being added using the old way

On Going
========
Scenario's that require more than one sample
Scenario's with multiple Collection Exercises
Graceful exit after test/system failures/exceptions
How to handle Scenario Outlines
Logging
Debugging multi-threaded app in Pycharm

# What has changed

New environment variables
=========================
IGNORE_SEQUENTIAL_DATA_SETUP = skips the 4 standard non-standalone collection exercises + respondent '49900000001' registration - defaults to 'False'

New tags
========
@business - currently does nothing but may be useful in future to separate business/social test runs
@standalone - enables a Scenario to run in parallel and can be used with the "@fixture" tag
@fixture.xxxxx - allows Scenario specific data setup

New files
=========
run_tags.py  - run tests by tag(s) - a temp solution (hopefully) to help debugging tests during development.
run_in_parallel - run standalone Scenarios concurrently 

# How to test?

Everything
==========
from a terminal in the rasrm-acceptance-tests directory, run - "make acceptance_tests" then "make acceptance_parallel_tests", all 211 should pass and 21 Surveys (name prefixed with "BUSINESS" or "SOCIAL") are on the "View list of business surveys" page 

Just Sequential (non-standalone) tests
============================
from a terminal in the rasrm-acceptance-tests directory, run - "make acceptance_tests", all 173 tests should pass

Just Parallel tests (no need to run "make setup" anymore :-) )
============================================
from a terminal in the rasrm-acceptance-tests directory, run - "make acceptance_parallel_tests", 38 tests should pass with 21 new surveys showing on the "View list of business surveys" pagepage 

# Links
[Test in conjunction with](https://github.com/ONSdigital/response-operations-ui/pull/248)
